quinn/src/lib.rs

```diff
@@ -798,93 +618,220 @@ impl Connection {
 
     /// The peer's UDP address.
     pub fn remote_address(&self) -> SocketAddr {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .connection(self.0.handle)
-            .remote()
+        self.0.borrow().inner.remote()
     }
 
-    /// The `ConnectionId`s defined for `conn` locally.
-    pub fn local_ids(&self) -> impl Iterator<Item = ConnectionId> {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .loc_cids(self.0.handle)
-            .cloned()
-            .collect::<Vec<_>>()
-            .into_iter()
-    }
     /// The `ConnectionId` defined for `conn` by the peer.
     pub fn remote_id(&self) -> ConnectionId {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .connection(self.0.handle)
-            .rem_cid()
+        self.0.borrow().inner.rem_cid()
     }
 
     /// The negotiated application protocol
     pub fn protocol(&self) -> Option<Box<[u8]>> {
-        self.0
-            .endpoint
-            .borrow()
-            .inner
-            .connection(self.0.handle)
-            .protocol()
-            .map(|x| x.into())
+        self.0.borrow().inner.protocol().map(|x| x.into())
     }
 
     // Update traffic keys spontaneously for testing purposes.
     #[doc(hidden)]
     pub fn force_key_update(&self) {
-        self.0
-            .endpoint
-            .borrow_mut()
-            .inner
-            .force_key_update(self.0.handle)
+        self.0.borrow_mut().inner.force_key_update()
     }
 }
 
 struct ConnectionInner {
-    endpoint: Rc<RefCell<EndpointInner>>,
+    log: Logger,
+    inner: quinn::Connection,
+    driver: Option<Task>,
     handle: ConnectionHandle,
+    pending: Pending,
     side: Side,
+    // TODO: Replace this with something custom that avoids using oneshots to cancel
+    timers: FuturesUnordered<Timer>,
+    conn_events: mpsc::UnboundedReceiver<ConnectionEvent>,
+    endpoint_events: mpsc::UnboundedSender<(ConnectionHandle, EndpointEvent)>,
+    connected: bool,
+    closed: bool,
```

Comment:
It's unclear if this is intended to express "closed gracefully" (in which case unbounded futures/streams could end without an error, but must also check pending.error to determine if the connection failed) or "closed for any reason".

Reply:
Right. I think in what I have now it's mostly used for "closed locally". I'll admit to having gotten a bit lost in the mechanics of it all, and I think we should come up with good invariants on what we want the externally visible behavior of the API want to be, and then we can figure out how to implement that.

But at the same time, this now passes all our tests and represents something like a minimal change to do so, so I'm not sure how much further change we should require to be part of this PR.